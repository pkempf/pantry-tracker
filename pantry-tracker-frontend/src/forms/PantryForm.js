import React, { useState } from "react";
import { LinkContainer } from "react-router-bootstrap";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Card from "react-bootstrap/Card";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";

const PantryForm = ({
  formTitle,
  fields,
  submitButtonText,
  processData,
  backButtonTo = "",
}) => {
  const getInitialState = (fields) => {
    let res = {};
    for (let i = 0; i < fields.length; i++) {
      res[fields[i].name] = fields[i].initialValue;
    }
    return res;
  };

  const [formData, setFormData] = useState(getInitialState(fields));

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value,
    }));
  };

  const getInput = (evt) => {
    evt.preventDefault();
    processData(formData);
  };

  return (
    <Row className="PantryForm justify-content-center">
      <Col sm={10} md={8}>
        <Card>
          <Card.Header as="h4">{formTitle}</Card.Header>
          <Card.Body>
            <Form onSubmit={getInput}>
              {fields.map((field) => {
                return (
                  <Form.Group key={field.name} controlId={field.name}>
                    <Form.Label>{field.label}</Form.Label>
                    <Form.Control
                      type={field.inputType}
                      name={field.name}
                      value={formData[field.name] || ""}
                      onChange={handleChange}
                      required={field.required}
                      placeholder={field.placeholder || ""}
                      readOnly={field.readOnly}
                      style={field.styleOverride || {}}
                    />
                  </Form.Group>
                );
              })}
              <Button variant="primary" type="submit" className="mr-2">
                {submitButtonText}
              </Button>
              {backButtonTo !== "" ? (
                <LinkContainer to={backButtonTo}>
                  <Button variant="secondary">Back</Button>
                </LinkContainer>
              ) : null}
            </Form>
          </Card.Body>
        </Card>
      </Col>
    </Row>
  );
};

export default PantryForm;
