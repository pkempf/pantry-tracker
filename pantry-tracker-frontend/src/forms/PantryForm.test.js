import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import PantryForm from "./PantryForm";

it("renders without crashing", () => {
  render(
    <MemoryRouter>
      <PantryForm fields={[]} />
    </MemoryRouter>
  );
});

it("matches the snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <PantryForm fields={[]} />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
