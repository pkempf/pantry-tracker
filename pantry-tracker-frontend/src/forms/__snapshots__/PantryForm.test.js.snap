// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="PantryForm justify-content-center row"
  >
    <div
      class="col-md-8 col-sm-10"
    >
      <div
        class="card"
      >
        <h4
          class="card-header"
        />
        <div
          class="card-body"
        >
          <form
            class=""
          >
            <button
              class="mr-2 btn btn-primary"
              type="submit"
            />
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
